src/stdlib/data.rs

```diff
@@ -0,0 +1,382 @@
+    filter_func,
+    |_, list, pred| {
+        expect_iterable!(list => iter);
+        expect_func!(pred => pred);
+        let next: Rc<RefCell<Option<FenderReference>>> = Rc::new(RefCell::new(None));
+        let next2 = next.clone();
+        let iter2 = iter;
+        let pred2 = pred.clone();
+        Ok(FenderIterator {
+            next: Rc::new(move |e| {
+                if let Some(val) = std::mem::take(&mut *next.borrow_mut()) {
+                    Ok(val)
+                } else {
+                    Ok(Default::default())
+                }
+            }),
+            has_next: Rc::new(move |e| {
+                if next2.borrow().is_some() {
+                    Ok(FenderValue::Bool(true).into())
+                } else {
+                    while *(iter2.has_next)(e)? == FenderValue::Bool(true) {
+                        let elem = (iter2.next)(e)?;
+                        if e.call(&pred2, vec![elem.clone()])?.to_bool() == FenderValue::Bool(true)
+                        {
+                            next2.replace(Some(elem));
+                            break;
+                        }
+                    }
+                    Ok(FenderValue::Bool(next2.borrow().is_some()).into())
+                }
+            }),
+        }
+        .into())
+    }
+);
+
+fndr_native_func!(
+    /// Reduce an iterator down to a single value with a starting value and accumulator
+    reduce_func,
+    |ctx, list, mut init, accum| {
+        expect_iterable!(list => iter);
+        expect_func!(accum => accum);
+        while *(iter.has_next)(ctx)? == FenderValue::Bool(true) {
+            let elem = (iter.next)(ctx)?;
+            init = ctx.call(accum, vec![init, elem])?;
+        }
+        Ok(init)
+    }
+);
+
+fndr_native_func!(
+    /// Reduce an iterator, but pass each intermediate reduction step forward
+    scan_func,
+    |ctx, list, mut init, accum| {
+        expect_iterable!(list => iter);
+        expect_func!(accum => accum);
+        let accum = accum.clone();
+        let init = RefCell::new(init);
+        Ok(FenderIterator {
+            next: Rc::new(move |e| {
+                let elem = (iter.next)(e)?;
+                let mut borrow = init.borrow_mut();
+                *borrow = e.call(&accum, vec![borrow.clone(), elem])?;
+                Ok(borrow.clone())
+            }),
+            has_next: iter.has_next.clone(),
+        }
+        .into())
+    }
+);
+
+fndr_native_func!(
+    /// Run a function on each element of an iterator
+    each_func,
+    |ctx, list, func| {
+        expect_iterable!(list => iter);
+        expect_func!(func => func);
+        while *(iter.has_next)(ctx)? == FenderValue::Bool(true) {
+            let elem = (iter.next)(ctx)?;
+            ctx.call(func, vec![elem])?;
+        }
+        Ok(Default::default())
+    }
+);
+
+fndr_native_func!(
+    /// Combines two iterables into a single iterator over pairs of their elements
+    zip_func,
+    |ctx, list1, list2| {
```

Comment:
can this take a variadic number of iters so 3 or more can be ziped at once and we don't need to have nested zip results the way we do in rust

Reply:
Can do